development/extensions/tutorial_notifications.rst

```diff
@@ -0,0 +1,955 @@
+When the item is a post, the parent is the topic.
+And so on, and so forth.
+If there is no parent for the item, you can set this to zero: ``return 0;``.
+
+.. code-block:: php
+
+   /**
+    * Get the id of the parent.
+    *
+    * @param array	$data			The type notification specific data
+    * @return int				Identifier of the parent
+    * @static
+    */
+   public static function get_item_parent_id($data)
+   {
+       return 0;
+   }
+
+find_users_for_notification
+---------------------------
+This function is responsible for finding the users that need to be notified.
+It should return an array with the user identifiers as keys and the notification methods as value.
+There are various helper functions that help you achieve the desired outcome.
+
+check_user_notification_options
++++++++++++++++++++++++++++++++
+You can send an array of user identifiers to this function.
+This function will then check the available notification methods for each user.
+If the notification type is available in the :abbr:`UCP (User Control Panel)`, it will check the user's preferences.
+Otherwise it will use the default notification methods; the board method and the email method (if *board-wide emails* is enabled).
+The array that is returned by this function, can be used as a return value for find_users_for_notification_.
+
+get_authorised_recipients
++++++++++++++++++++++++++
+If your notification is for an event within a specific forum, you might want to check the users' authentication.
+This can be done by using this function, which will check all users' ``f_read`` permission for the provided ``forum_id``.
+The array that is returned by this function is already put through check_user_notification_options_.
+
+.. code-block:: php
+
+   /**
+    * Find the users who want to receive notifications.
+    *
+    * @param array $data			The type specific data
+    * @param array $options			Options for finding users for notification
+    * 				 		ignore_users => array of users and user types that should not receive notifications from this type
+    * 				 		because they've already been notified
+    * 						  e.g.: array(2 => array(''), 3 => array('', 'email'), ...)
+    * @return array				Array of user identifiers with their notification method(s)
+    */
+   public function find_users_for_notification($data, $options = [])
+   {
+       return $this->check_user_notification_options([$data['user_id']], $options);
+   }
+
+.. note::
+
+    The ``forum_id`` variable below is not set in this Tutorial, just shown as an example.
+
+.. code-block:: php
+
+   public function_find_users_for_notification($data, $options = [])
+   {
+       return $this->get_authorised_recipients([$data['user_id']], $data['forum_id'], $options);
+   }
+
+users_to_query
+--------------
+This function should return an array of user identifiers.
+Identifiers of users that need to be queried for this notification to be displayed to the user that is notified.
```

Comment:
I'm finding this text difficult to understand, for a function I always found difficult to understand too, so hopefully this can be better explained.